Sources/MongoSwift/MongoClient.swift

```diff
@@ -322,11 +317,33 @@ public class MongoClient {
      */
     public func withSession<T>(
         options: ClientSessionOptions? = nil,
-        _ sessionBody: (ClientSession) throws -> T
-    ) throws -> T {
-        let session = try ClientSession(client: self, options: options)
-        defer { session.end() }
-        return try sessionBody(session)
+        _ sessionBody: (ClientSession) throws -> EventLoopFuture<T>
+    ) -> EventLoopFuture<T> {
+        let promise = self.operationExecutor.makePromise(of: T.self)
+        let session = self.startSession(options: options)
+        do {
+            let bodyFuture = try sessionBody(session)
+            // regardless of whether body's returned future succeeds we want to call session.end() once its complete.
+            // only once session.end() finishes can we fulfill the returned promise. otherwise the user can't tell if
+            // it is safe to close the parent client of this session, and they could inadvertently close it before the
+            // session is actually ended and its parent `mongoc_client_t` is returned to the pool.
+            bodyFuture
+                .flatMap { _ in
+                    session.end()
+                }
+                .flatMapError { _ in
```

Comment:
If you remove the `multiline_function_chains` rule from the swiftlint config (as I have done in the cursors PR), then these chains can be on the same line

Reply:
done